### DISCUSSION TITLE
Developing via native MacOS, is it possible?

### DISCUSSION THREAD

**Discussion by forsamori** [mastodon/mastodon#25441] [Q&A] [answered]
Hi folks,

Is it possible to develop for Mastodon (i.e contribute to Mastodon source) natively on MacOS? 

I think the only thing inside the setup steps that isn’t possible (at least easily, to my knowledge) on Mac is `apt`. 

If it’s absolutely not possible, I’ll just spin up an Ubuntu 20.04 VM - I’m just more comfortable natively. 

Thanks!

**Comment by renchap** [comment, marked as answer]
Yes, you can do it without any specific issue.

You need:
- a working Ruby 3.2 install (you can use https://github.com/rbenv/rbenv to easily install and manage ruby versions)
- postgresql (`brew install postgresql@15`)
- redis (`brew install redis`)
- imagemagick (`brew install imagemagick`)
- Foreman or a similar tool to launch multiple processes using the `Procfile.dev` file, I use https://github.com/DarthSim/overmind
- Node.js version 16 (18+ does not work yet). Easiest way is to use nvm (`brew install nvm` then run `nvm use` in Mastodon's root directory to have the correct version installed)
- yarn: `corepack enable && yarn set version classic`

I think that should be all, you should now be able to run `bundle exec rails db:setup`, then `overmind start -f Procfile.dev` to start the various processes.

**Reply by forsamori** [reply]
Amazing, thank you :) I’ve got most of those already from my day-to-day. Looking forward to picking up some bugs to have a shot at :)

**Reply by forsamori** [reply]
@renchap So it looks like your solution here will work fine, so long as you're not using an M1/2 Mac. Unfortunately, VirtualBox is not fully supported yet on M1/2, so VMs fail to launch. I'm looking at alternative VM providers to see if I can maybe update my local Vagrant to run on something like UTM.

**Reply by renchap** [reply]
You dont need Vagrant for this, I am running everything locally on an M1 mac without any issues.

**Reply by forsamori** [reply]
Yep, that's me making progress now, thanks!

**Reply by forsamori** [reply]
Setup complete, thanks @renchap

**Comment by varun-s22** [comment]
i also have a mac, but sometimes people do not wish to install dependencies like redis or ruby in their system.
Maybe mastodon should have a docker file for running everything inside a container(in dev), to avoid installation of dependencies on developer's desktop.
